### PR TITLE
[JBS-328] [JBS-330] Add more metrics for cinder and ec2api layer. Making Metrics more generic and moving some common utility packages here

### PR DESCRIPTION
**Pull request by souvik-ray**
Already tested the code in cinder ec2 layer.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,4 +5,4 @@
 build/
 dist/
 metrics.egg-info/
-
+.idea/
```
Comment (vivekdhayaal) [inline .gitignore]: unintended change?

File: metrics/AbstractMetrics.py [hunk 2]
```diff
@@ -148,8 +148,11 @@ def __isTimingMetric(self, unit):
 class AbstractMetricsFactory(MetricsFactory):
     ''' An abstract factory which creates some basic attributes of a metric object
     '''
-    
-    __metaclass__ = ABCMeta
```
Comment (vivekdhayaal) [inline metrics/AbstractMetrics.py]: I believe you dont AbstractMetricsFactory to be instantiable. Then why this change?

File: metrics/AbstractMetrics.py [hunk 3]
```diff
@@ -148,8 +148,11 @@ def __isTimingMetric(self, unit):
...
+    # TODO: This classes should not be instantiable.
+    # But we need to solve this problem first.
+    # metaclass conflict: the metaclass of a derived class must be a (non-strict)
```
Comment (vivekdhayaal) [inline metrics/AbstractMetrics.py]: "the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases". Can you explain what is your expectation here?

File: metrics/Metrics.py [hunk 4]
```diff
@@ -158,7 +158,11 @@ def close(self):
 class MetricsFactory(object):
     """ This is an interface factory method to create an instance of Metrics class
     """
-    __metaclass__ = ABCMeta
+    # TODO: This classes should not be instantiable.
+    # But we need to solve this problem first.
+    # metaclass conflict: the metaclass of a derived class must be a (non-strict)
```
Comment (vivekdhayaal) [inline metrics/Metrics.py]: Same here.

File: metrics/metric_util.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
+'''
```
Comment (vivekdhayaal) [inline metrics/metric_util.py]: This code is already part of cinder. 
https://github.com/JioCloud/cinder/commit/0a4b74156c5e0024c2a761e140e2fd78eb780315
Why duplicate here again?

Will review this file after your answer...

File: metrics/ThreadLocalMetrics.py [hunk 6]
```diff
@@ -58,18 +43,50 @@ def close(self):
...
+    #The default path is given to ensure backward compatibility to cinder. To be removed soon.
+    def create_timed_rotating_log(self, path = "/var/log/cinder/service.log", propagate_to_application_logs = True):
+        ''' This method describes the logging type of service logs
+        '''
+        logger = logging.getLogger("service.log")
+        logger.propagate = propagate_to_application_logs
+        # Uncomment this after thorough validation in Production, that all metrics are in service
+        # logs. This will remove the service logs from cinder API.
+
+        handler = logging.handlers.WatchedFileHandler(path)
```
Comment (vivekdhayaal) [inline metrics/ThreadLocalMetrics.py]: This would break, if the path doesn't exist. Atleast that's what happened when I ran the unit tests. Check the patch I'd done @ 
https://jira.ril.com/browse/JBS-214?focusedCommentId=164770&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-164770

File: metrics/ThreadLocalMetrics.py [hunk 7]
```diff
@@ -58,18 +43,50 @@ def close(self):
...
+    '''
     def create_metrics(self):
-
-        metrics = ThreadLocalMetrics()
+        metrics = ThreadLocalMetrics(self.__logger)
         self._add_metric_attributes(metrics)
         return metrics
+
+    #The default path is given to ensure backward compatibility to cinder. To be removed soon.
+    def create_timed_rotating_log(self, path = "/var/log/cinder/service.log", propagate_to_application_logs = True):
```
Comment (vivekdhayaal) [inline metrics/ThreadLocalMetrics.py]: the filename is hardcoded in multiple places. the cleaner way is to declare a broader scoped string constant.

As said earlier, "propagate_to_application_logs = False" ?

File: metrics/ThreadLocalMetrics.py [hunk 8]
```diff
@@ -58,18 +43,50 @@ def close(self):
...
+    metricsFactory = ThreadLocalMetricsFactory("/tmp/service_log").with_marketplace_id("IDC1").with_program_name("CinderAPI")
     metrics =  metricsFactory.create_metrics();
     '''
+    __metaclass__ = Singleton
 
-    def __init__(self, service_log_path):
+    def __init__(self, service_log_path, propagate_to_application_logs = True):
         super(ThreadLocalMetricsFactory, self).__init__()
-
+        # This is done so that metrics flowing in cinder do not break.  To be removed soon.
```
Comment (vivekdhayaal) [inline metrics/ThreadLocalMetrics.py]: Need some more explanation to understand what the next line does.

File: metrics/ThreadLocalMetrics.py [hunk 9]
```diff
@@ -58,18 +43,50 @@ def close(self):
...
     ''' Factory method to create Thread Local Metrics
     Example Usage:
-    metricsFactory = ThreadLocalMetricsFactory("/tmp/service_log").with_account_id("xxxxxxxxxxxxxxx").with_marketplace_id("IDC1").with_program_name("CinderAPI").with_operation_name("CreateVolume");
+    metricsFactory = ThreadLocalMetricsFactory("/tmp/service_log").with_marketplace_id("IDC1").with_program_name("CinderAPI")
     metrics =  metricsFactory.create_metrics();
     '''
+    __metaclass__ = Singleton
 
-    def __init__(self, service_log_path):
+    def __init__(self, service_log_path, propagate_to_application_logs = True):
```
Comment (vivekdhayaal) [inline metrics/ThreadLocalMetrics.py]: I believe, ultimately we dont want duplication of metrics logs in cinder logs. So, shouldn't this be "False" and overridden with a "True" while instantiating, with an explicit code comment on the rationale?

File: metrics/ThreadLocalMetrics.py [hunk 10]
```diff
@@ -58,18 +43,50 @@ def close(self):
...
+    def __call__(cls, *args, **kwargs):
+        if cls.instance is None:
+            cls.instance = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls.instance
+
 class ThreadLocalMetricsFactory(AbstractMetricsFactory):
     ''' Factory method to create Thread Local Metrics
     Example Usage:
-    metricsFactory = ThreadLocalMetricsFactory("/tmp/service_log").with_account_id("xxxxxxxxxxxxxxx").with_marketplace_id("IDC1").with_program_name("CinderAPI").with_operation_name("CreateVolume");
+    metricsFactory = ThreadLocalMetricsFactory("/tmp/service_log").with_marketplace_id("IDC1").with_program_name("CinderAPI")
```
Comment (vivekdhayaal) [inline metrics/ThreadLocalMetrics.py]: Why "/tmp"?

File: metrics/ThreadLocalMetrics.py [hunk 11]
```diff
@@ -10,33 +9,19 @@
 from AbstractMetrics import AbstractMetrics, AbstractMetricsFactory
 import threading
 
-def create_timed_rotating_log(path):
```
Comment (vivekdhayaal) [inline metrics/ThreadLocalMetrics.py]: why do u want to move it from global to class scope?

File: .gitignore (hunk 1 above)
Comment (souvik-ray) [inline .gitignore]: Totally intended.. 
as discussed i want this to be ignored by gitignore

File: metrics/AbstractMetrics.py (hunk 3 above)
Comment (vivekdhayaal) [inline metrics/AbstractMetrics.py]: As discussed offline, your expectation is that AbstractMetricsFactory should be an abstract and ThreadLocalMetricsFactory should be a singleton. To achieve this behaviour without a metaclass conflict, do this:

class TLMFMeta(ABCMeta, Singleton):
    pass

class ThreadLocalMetrics(...):
    ...
    __metaclass__ = TLMFMeta

File: metrics/metric_util.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    def __call__(self, function):
+        def metrics_wrapper(*args, **kwargs):
+            start_time = time()
+            error = 0
+            try:
+                return function(*args, **kwargs)
+            except Exception as e:
+                LOG.error("Exception while executing " + function.__name__)
+                error = 1
+                raise e
```
Comment (vivekdhayaal) [inline metrics/metric_util.py]: Could you please update the test logs here? because I'm not sure if this really works. This morning, when I was working on JBS-342, I found that the actual exception trace wasn't printed at this point. Please check the below snip.

<snip>
...
2016-05-06 10:57:11.861 31654 INFO cinder.api.openstack.wsgi [req-7121aafb-2ec9-4af2-adf1-c305150f93dc 08f5800c1b4f42f39349228a57983f51 00000000000000000000306638138824 - - -] POST http://10.140.211.140:81/v2/00000000000000000000306638138824/backups
2016-05-06 10:57:11.863 31654 DEBUG cinder.api.contrib.backups [req-7121aafb-2ec9-4af2-adf1-c305150f93dc 08f5800c1b4f42f39349228a57983f51 00000000000000000000306638138824 - - -] Creating new backup {u'backup': {u'volume_id': u'0ad650d1-ae47-4a7e-86a7-7bb7a63b84c1', u'container': None, u'description': None, u'name': None}} create /usr/lib/python2.7/dist-packages/cinder/api/contrib/backups.py:251
2016-05-06 10:57:11.864 31654 INFO cinder.api.contrib.backups [req-7121aafb-2ec9-4af2-adf1-c305150f93dc 08f5800c1b4f42f39349228a57983f51 00000000000000000000306638138824 - - -] Creating backup of volume 0ad650d1-ae47-4a7e-86a7-7bb7a63b84c1 in container None
2016-05-06 10:57:11.928 31654 DEBUG cinder.quota [req-fed38de0-ed07-469f-829c-2850a2a6318d - - - - -] Created reservations ['6302b5c1-d6a3-48b7-be6e-1923110e230d', 'a9203210-4f38-42f5-a105-4b683169ba09'] reserve /usr/lib/python2.7/dist-packages/cinder/quota.py:767
2016-05-06 10:57:11.983 31654 ERROR cinder.api.metricutil [req-fed38de0-ed07-469f-829c-2850a2a6318d - - - - -] Exception while executing create
2016-05-06 10:57:11.984 31654 INFO cinder.api.openstack.wsgi [req-fed38de0-ed07-469f-829c-2850a2a6318d - - - - -] http://10.140.211.140:81/v2/00000000000000000000306638138824/backups returned with HTTP 400
2016-05-06 10:57:11.984 31654 INFO service.log [req-fed38de0-ed07-469f-829c-2850a2a6318d - - - - -] ===================================================================
StartTime=1462512431.86
EndTime=2016-05-06 10:57:11
Time=121 ms
OperationName=create
RemoteAddress=10.140.211.131
PathInfo=/00000000000000000000306638138824/backups
HostName=sbsep1
PID=31654
TenantId=00000000000000000000306638138824
RequestId=req-7121aafb-2ec9-4af2-adf1-c305150f93dc
ProgramName=CinderAPI
MarketplaceId=IDC1
Timing=backup-api-create_time=118 ms,mysql-db-volume-get_time=38 ms,mysql-db-backup-getall-volume_time=52 ms,volume-api-get_time=38 ms,
Counters=fault=0,success=1,error=0,
EOE
2016-05-06 10:57:11.986 31654 INFO eventlet.wsgi.server [req-fed38de0-ed07-469f-829c-2850a2a6318d - - - - -] 10.140.211.131 - - [06/May/2016 10:57:11] "POST /v2/00000000000000000000306638138824/backups HTTP/1.1" 400 342 0.204303
...
</snip>

File: metrics/metric_util.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    '''
+    Metric Utility class to put and fetch request scoped metrics in cinder api
+    '''
+    METRICS_OBJECT = "metrics_object"
+    def __init__(self):
+        '''
+        Constructor for Metric Utils. 
+        '''
+
+    '''
```
Comment (vivekdhayaal) [inline metrics/metric_util.py]: Nice-to-have: Docs should be nested inside class/method definitions

File: metrics/ThreadLocalMetrics.py (hunk 9 above)
Comment (vivekdhayaal) [inline metrics/ThreadLocalMetrics.py]: As discussed, it would be nice to record your offline justification here...

File: metrics/metric_util.py (hunk 5 above)
Comment (vivekdhayaal) [inline metrics/metric_util.py]: As discussed, it would be nice to record your offline justification here...

File: metrics/ThreadLocalMetrics.py (hunk 11 above)
Comment (vivekdhayaal) [inline metrics/ThreadLocalMetrics.py]: As discussed, it would be nice to record your offline justification here...

**Comment by vivekdhayaal** [comment]
Two things:
1)Update the commit message(git commit --amend) with the ID of the JIRA ticket you are fixing.
2)As is the usual practice, update the logs of your testing in JIRA or here

File: metrics/AbstractMetrics.py (hunk 3 above)
Comment (souvik-ray) [inline metrics/AbstractMetrics.py]: This is great .....
Hats off to your python knowledge :)

File: metrics/ThreadLocalMetrics.py (hunk 6 above)
Comment (souvik-ray) [inline metrics/ThreadLocalMetrics.py]: As said this is for backward compatibility of cinder .... (This is more important here)
Once this code is merged and pushed I will clean this up..

Till then I will ignore the unit-test results...

File: metrics/ThreadLocalMetrics.py (hunk 7 above)
Comment (souvik-ray) [inline metrics/ThreadLocalMetrics.py]: Again as discussed offline Propagating is necessary because of the multiple request Id bug...

File: metrics/ThreadLocalMetrics.py (hunk 8 above)
Comment (souvik-ray) [inline metrics/ThreadLocalMetrics.py]: Same backward compatibility

File: metrics/ThreadLocalMetrics.py (hunk 10 above)
Comment (souvik-ray) [inline metrics/ThreadLocalMetrics.py]: Its an example :)

File: metrics/metric_util.py (hunk 5 above)
Comment (souvik-ray) [inline metrics/metric_util.py]: 1) Today keystone calls are not within purview of metrics.. This change will do so
2)  It will have better fault results

File: metrics/ThreadLocalMetrics.py (hunk 11 above)
Comment (souvik-ray) [inline metrics/ThreadLocalMetrics.py]: This is now configurable and multiple services can use it....
Global configuration does not allow you to do so...

Singleton ensures that this does not get instantiated everytime...

File: metrics/metric_util.py (hunk 13 above)
Comment (souvik-ray) [inline metrics/metric_util.py]: Please take a class on python docs... You surely have great knowledge here .. :)

File: metrics/ThreadLocalMetrics.py (hunk 6 above)
Comment (vivekdhayaal) [inline metrics/ThreadLocalMetrics.py]: Its ok to ignore unit-tests but my concern was not that. unit tests was just an instance. The concern is...will cinder/metrics work if the file doesn't exist? My suspect is that it will fail if the file doesn't exist. It would be good to test this scenario and update the results here.

Comment (souvik-ray) [inline metrics/ThreadLocalMetrics.py]: No this is a logger file and this creates automatically.... (atleast as long as /var/log/cinder exists) 
Since this is only used in cinder or ec2 api layer where the log folder already exists ....
I am not too worried about any other use cases which is not relevant now

No body created this file in production/staging and it has been TESTED...........

File: metrics/metric_util.py (hunk 12 above)
Comment (souvik-ray) [inline metrics/metric_util.py]: the exception there was a Volume not found exception.....
2016-05-06 10:57:11.984 31654 INFO cinder.api.openstack.wsgi [req-fed38de0-ed07-469f-829c-2850a2a6318d - - - - -] http://10.140.211.140:81/v2/00000000000000000000306638138824/backups returned with HTTP 400

The mysql layer had thrown the error but on a volume not found exception.. and since mysql error is thrown on not found the metrics wrapper thinks this is an error....
Will think of ways to clean up ... but as of now i ll keep it like this... as you can see what the issue is...

Comment (vivekdhayaal) [inline metrics/metric_util.py]: Carried forward to JBS-358

File: metrics/ThreadLocalMetrics.py (hunk 8 above)
Comment (vivekdhayaal) [inline metrics/ThreadLocalMetrics.py]: Carried forward to JBS-358

File: metrics/ThreadLocalMetrics.py (hunk 7 above)
Comment (vivekdhayaal) [inline metrics/ThreadLocalMetrics.py]: Carried forward to JBS-358

File: metrics/ThreadLocalMetrics.py (hunk 6 above)
Comment (vivekdhayaal) [inline metrics/ThreadLocalMetrics.py]: Carried forward to JBS-358

**Comment by vivekdhayaal** [comment]
Souvik wants this to be merged today and be ready for deployment tomorrow; Merged this for now as the outstanding issues are tracked @ 
https://jira.ril.com/browse/JBS-358
